File: 739. Daily Temperatures/Solution.py
class Solution(object):
    def dailyTemperatures(self, T):
        """
        :type T: List[int]
        :rtype: List[int]
        """
        nxt = [float("inf") for _ in range(102)]
        ans = [0] * len(T)
        
        for i in range(len(T) - 1, -1, -1):
            idx = min(nxt[t] for t in range(T[i] + 1, 102))
            if idx < float("inf"):
                ans[i] = idx - i
            nxt[T[i]] = i
        return ans

# stack
class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        res = [0 for _ in range(len(T))]
        stack = []
        
        for i in range((len(T) - 1), -1, -1):
            while stack and T[stack[-1]] <= T[i]:
                stack.pop()
            if stack == []:
                res[i] = 0
            else:
                res[i] = stack[-1] - i
            stack.append(i)
        
        return res

File: 739. Daily Temperatures/Solution.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] index = new int[n];
        Stack<Integer> stack = new Stack<>();
        
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && temperatures[i] >= temperatures[stack.peek()]) {
                stack.pop();
            }
            index[i] = stack.isEmpty() ? 0 : (stack.peek() - i);
            stack.push(i);
        }
        
        return index;
    }
}